### PR TITLE
fix SslStream.IsMutuallyAuthenticated with cached credentials

### LINKED ISSUE #65563: SslStream.IsMutuallyAuthenticated returns wrong value when cached creds are used
### Description

We use `SslStream` to connect with a device which allows to be configured for mutual authentication with client certificates. For that we use `AuthenticateAsClientAsync`. After calling this method we check the `IsMutuallyAuthenticated` property the way it is mentioned in the docs:

> ...Check the [IsMutuallyAuthenticated](https://docs.microsoft.com/en-us/dotnet/api/system.net.security.sslstream.ismutuallyauthenticated?view=net-6.0) property to determine whether mutual authentication occurred.

see [Remarks section](https://docs.microsoft.com/en-us/dotnet/api/system.net.security.sslstream.authenticateasclientasync?view=net-6.0#system-net-security-sslstream-authenticateasclientasync(system-string-system-security-cryptography-x509certificates-x509certificatecollection-system-security-authentication-sslprotocols-system-boolean)) 

This works fine so far. But subsequent connect with authenticate to the same device (server) now results in that the property `IsMutuallyAuthenticated` returns `true` even when this is not correct anymore.

This wrong behavior comes from the caching which is done by the SslStream as it is mentioned in the docs:

> Note:
The Framework caches SSL sessions as they are created and attempts to reuse a cached session for a new request, if possible. When attempting to reuse an SSL session, the Framework uses the first element of [ClientCertificates](https://docs.microsoft.com/en-us/dotnet/api/system.net.httpwebrequest.clientcertificates?view=net-6.0) (if there is one), or tries to reuse an anonymous sessions if [ClientCertificates](https://docs.microsoft.com/en-us/dotnet/api/system.net.httpwebrequest.clientcertificates?view=net-6.0) is empty.

see [Remarks section](https://docs.microsoft.com/en-us/dotnet/api/system.net.security.sslstream.-ctor?view=net-6.0#system-net-security-sslstream-ctor(system-io-stream-system-boolean-system-net-security-remotecertificatevalidationcallback))

### Reproduction Steps

1. Connect to a server via SslStream which requires mutual authentication via client certificate.
2. Call `AuthenticateAsClientAsync` and provide the client certificate
3. Verify that `IsMutuallyAuthenticated` returns true
4. Disconnect

- Note: Keep the client running

5. Restart and reconfigure the server so that it does not require a mutual authentication anymore
6. Connect to a server via SslStream
7. Call `AuthenticateAsClientAsync` and provide the client certificate
8. Verify what `IsMutuallyAuthenticated` returns now (see Expected and Actual behavior)

### Expected behavior

In step 8:
8. `IsMutuallyAuthenticated` should return false

### Actual behavior

In step 8:
8. `IsMutuallyAuthenticated` returns true

Although a TLS 1.2 connection was established without mutual authentication.

### Regression?

I can reproduce this issue with:
- .NET 6
- .NET 5
- .NET Framework 4.8

### Known Workarounds

This not-recommended workaround solves the issue:

- Call the following method before starting connecting to the server.

```C#
private static void ClearCache()
{
    var sslAssembly = Assembly.GetAssembly(typeof(SslStream));
    var sslSessionCacheClass = sslAssembly.GetType("System.Net.Security.SslSessionsCache");
    var cachedCredsInfo = sslSessionCacheClass.GetField("s_cachedCreds", BindingFlags.NonPublic | BindingFlags.Static);
    var cachedCreds = cachedCredsInfo.GetValue(null);
    cachedCreds.GetType().GetMethod("Clear", BindingFlags.Public | BindingFlags.Instance).Invoke(cachedCreds, null);
}
```

### Configuration

- Windows 10 20H2

### Other information

Maybe this is related: https://github.com/dotnet/runtime/issues/65539

### PR DESCRIPTION
**Pull request by wfurt**
fixes #65563

We can query Schannel after handshake is done to see if certificate was used. This is no-op for test of the platforms at the moment.
The test is essentially repro @rzikm put together during investigation. 

not 100% about android. cc: @simonrozsival

### CODE REVIEW

**Review by simonrozsival — COMMENTED** [review]
I'm not very familiar with the client selection code on Android and it's not very easy to follow, but this is my current interpretation of the what we do:
- whenever there's a `CertificateContext` in the auth options, we initialize the Java KeyStore with the client certificate. The Java SSLEngine may or may not send the client certificate to the server. We never call the `SelectClientCertificate`. We don't check if the client sent the certificate to the server or not.
- the `CertificateContext` is set by the `AcquireClientCredentials` method which calls `SelectClientCertificate` method that sets the `_selectedClientCertificate` field.
- So from what I can tell, on Android the `IsMutuallyAuthenticated` property could be `true` even though we never sent the client certificate to the server.

But... don't we call `AcquireClientCredentials` on all platforms? Isn't that the same case also on mac/iOS?

I think we should at least run tests on Android for this PR and see if the new test fails or not. I can do some more digging if necessary. It should be fairly easy to implement `IsLocalCertificateUsed` to actually check which certificate we sent to the server (using [SSLSession#getLocalCertificates()](https://developer.android.com/reference/javax/net/ssl/SSLSession#getLocalCertificates())).

**Comment by rzikm** [comment]
https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-79128-merge-1d0bff68b72b460f94/System.Net.Security.Tests/3/console.e5e0dbcd.log?helixlogtype=result

```
 System.Net.Security.Tests.SslStreamMutualAuthenticationTest.SslStream_CachedCredentials_IsMutuallyAuthenticatedCorrect(protocol: Tls12) [FAIL]
      Assert.Equal() Failure
      Expected: True
      Actual:   False
      Stack Trace:
        /_/src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamMutualAuthenticationTest.cs(121,0): at System.Net.Security.Tests.SslStreamMutualAuthenticationTest.SslStream_CachedCredentials_IsMutuallyAuthenticatedCorrect(SslProtocols protocol)
        --- End of stack trace from previous location ---
```

Looks like we fail on Win7

File: src/libraries/Common/src/Interop/Windows/SspiCli/SSPIWrapper.cs [hunk 1]
```diff
@@ -298,6 +298,10 @@ private static bool QueryCertContextAttribute(ISSPIInterface secModule, SafeDele
         public static bool QueryContextAttributes_SECPKG_ATTR_REMOTE_CERT_CONTEXT(ISSPIInterface secModule, SafeDeleteContext securityContext, out SafeFreeCertContext? certContext)
             => QueryCertContextAttribute(secModule, securityContext, Interop.SspiCli.ContextAttribute.SECPKG_ATTR_REMOTE_CERT_CONTEXT, out certContext);
 
+        public static bool QueryContextAttributes_SECPKG_ATTR_LOCAL_CERT_CONTEXT(ISSPIInterface secModule, SafeDeleteContext securityContext, out SafeFreeCertContext? certContext)
+            => QueryCertContextAttribute(secModule, securityContext, Interop.SspiCli.ContextAttribute.SECPKG_ATTR_LOCAL_CERT_CONTEXT, out certContext);
+
+
```
Comment (rzikm) [inline src/libraries/Common/src/Interop/Windows/SspiCli/SSPIWrapper.cs:304]:
```suggestion
```

File: src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.Windows.cs [hunk 2]
```diff
@@ -89,6 +89,28 @@ internal static partial class CertificateValidationPal
...
+                {
+                    return !localContext.IsInvalid;
+                }
+            }
+            finally
+            {
+                localContext?.Dispose();
+            }
+
+            // Some older Windows do not support that. This is only called when client certificate was provided
```
Comment (rzikm) [inline src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.Windows.cs:109]: Do unsupported platforms throw in this case? If so, won't that cause a perf regression?

**Comment by simonrozsival** [comment]
@wfurt The test is failing on Android. I have a fix ready so you can skip the test on Android and I'll open a follow-up PR once this one is merged.

**Comment by wfurt** [comment]
One day I'll learn how to run tests on Android locally ... but it may not be this PR.

File: src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.Windows.cs (hunk 2 above)
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.Windows.cs:109]: no, I don't think it throws. The Windows7 fails to even pass the first `IsMutuallyAuthenticate` Assert. The explanation is in test above and that test is also disabled on Windows7.  So I also disabled new test on Win7 (and it is about of support anyway)

**Comment by wfurt** [comment]
I disabled the test on Android with active issue pointing at original one. I'll keep it open until you are ready with the fix. 
(and thanks for looking into it)

**Comment by wfurt** [comment]
Last test (CertificateValidationClientServer_EndToEnd_Ok) was failing on server 2012. I investigated and the call would finish with `SEC_E_NO_CREDENTIALS` if the connection was resumed. That seems like issue with older schannel versions - never windows return the certificate status according to the previous. 

For now, I updated the test avoid this case. I'm not sure what we can do about it. I feel it would be better if we report the status correctly on current Windows, hopefully 2012 will go out of support soon. 
Let me know if you have any suggestions @rzikm.

**Review by rzikm — APPROVED** [review]

**Pull request merged**